ports/atmel-samd/timers.c

```diff
@@ -0,0 +1,285 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "timers.h"
+
+#include "common-hal/pulseio/PulseOut.h"
+
+#ifdef SAMD21
+#include "hpl/gclk/hpl_gclk_base.h"
+#endif
+
+#ifdef SAMD51
+#include "hri/hri_gclk_d51.h"
+#endif
+
+// This bitmask keeps track of which channels of a TCC are currently claimed.
+#ifdef SAMD21
+const uint8_t tcc_cc_num[3] = {4, 2, 2};
+const uint8_t tc_gclk_ids[TC_INST_NUM] = {TC3_GCLK_ID,
+               TC4_GCLK_ID,
+               TC5_GCLK_ID,
+#ifdef TC6_GCLK_ID
+             , TC6_GCLK_ID
+#endif
+#ifdef TC7_GCLK_ID
+             , TC7_GCLK_ID
+#endif
+            };
+const uint8_t tcc_gclk_ids[3] = {TCC0_GCLK_ID, TCC1_GCLK_ID, TCC2_GCLK_ID};
+#endif
+#ifdef SAMD51
+const uint8_t tcc_cc_num[5] = {6, 4, 3, 2, 2};
+const uint8_t tc_gclk_ids[TC_INST_NUM] = {TC0_GCLK_ID,
+                                          TC1_GCLK_ID,
+                                          TC2_GCLK_ID,
+                                          TC3_GCLK_ID
+#ifdef TC4_GCLK_ID
+                                          , TC4_GCLK_ID
+#endif
+#ifdef TC5_GCLK_ID
+                                          , TC5_GCLK_ID
+#endif
+#ifdef TC6_GCLK_ID
+                                          , TC6_GCLK_ID
+#endif
+#ifdef TC7_GCLK_ID
+                                          , TC7_GCLK_ID
+#endif
+                                      };
+const uint8_t tcc_gclk_ids[5] = {TCC0_GCLK_ID, TCC1_GCLK_ID, TCC2_GCLK_ID, TCC3_GCLK_ID,
+                                 TCC4_GCLK_ID};
+#endif
+Tc* const tc_insts[TC_INST_NUM] = TC_INSTS;
+Tcc* const tcc_insts[TCC_INST_NUM] = TCC_INSTS;
+
+IRQn_Type const tc_irq[TC_INST_NUM] = {
+#ifdef TC0
+    TC0_IRQn,
+#endif
+#ifdef TC1
+    TC1_IRQn,
+#endif
+#ifdef TC2
+    TC2_IRQn,
+#endif
+    TC3_IRQn,
+    TC4_IRQn,
+    TC5_IRQn
+#ifdef TC6
+    , TC6_IRQn
+#endif
+#ifdef TC7
+    , TC7_IRQn
+#endif
+};
+
+void turn_on_clocks(bool is_tc, uint8_t index, uint32_t gclk_index) {
+    uint8_t gclk_id;
+    if (is_tc) {
+        gclk_id = tc_gclk_ids[index];
+    } else {
+        gclk_id = tcc_gclk_ids[index];
+    }
+    // Turn on the clocks for the peripherals.
+    #ifdef SAMD51
+    if (is_tc) {
+        switch (index) {
```

Comment:
could be just a table lookup

Reply:
This is a little trickier because the register its assigning to changes. Its hard to see but the fourth letter changes. Still want me to have two look up tables?

Reply:
no, didn't see that!